### PR TITLE
Add Axom spack package

### PR DESCRIPTION
**Pull request by white238**
This adds an Axom spack package that runs on all the Axom supported LLNL systems.  I have run flake8 on all the files.

I am currently re-running our battery of tests which should pass because these files should have no changes other than style in them from what was already in Axom.  This takes a couple hours.

### CODE REVIEW

**Comment by white238** [comment]
This passes all internal Axom tests.

**Comment by white238** [comment]
Added missing shroud package that was already in Axom.

File: var/spack/repos/builtin/packages/axom/package.py [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+class Axom(Package):
+    """Axom provides a robust, flexible software infrastructure for the development
+       of multi-physics applications and computational tools."""
+
+    maintainers = ['white238']
+
+    homepage = "https://github.com/LLNL/axom"
+    git      = "https://github.com/LLNL/axom.git"
+
+    version('master', branch='master', submodules=True, preferred=True)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/axom/package.py]: We usually like the newest stable release to be the default version, as master is a changing tip that isn't reproducible.

File: var/spack/repos/builtin/packages/axom/package.py [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+            # Are we on a LLNL system then strip node number
+            hostname = hostname.rstrip('1234567890')
+        filename = "{0}-{1}-{2}.cmake".format(hostname,
+                                              self._get_sys_type(spec),
+                                              spec.compiler)
+        dest_dir = self.stage.source_path
+        fullpath = os.path.abspath(pjoin(dest_dir, filename))
+        return fullpath
+
+    def hostconfig(self, spec, prefix):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/axom/package.py:159]: Can these variables not be passed through `cmake_args`?

File: var/spack/repos/builtin/packages/raja/package.py [hunk 3]
```diff
@@ -15,7 +15,7 @@ class Raja(CMakePackage):
     version('develop', branch='develop', submodules='True')
     version('master',  branch='master',  submodules='True')
     version('0.11.0', tag='v0.11.0', submodules="True")
-    version('0.10.0', tag='v0.10.0', submodules="True")
+    version('0.10.1', tag='v0.10.1', submodules="True")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/raja/package.py:18]: Why is this change necessary?

File: var/spack/repos/builtin/packages/axom/package.py (hunk 2 above)
Comment (cyrush) [reply var/spack/repos/builtin/packages/axom/package.py:159]: a generated host config is our source of provenance and also a way to config again outside of spack

**Comment by cyrush** [comment]
conduit changes LGTM (Look good to me or Look galant to minotaur, either way good!).

File: var/spack/repos/builtin/packages/axom/package.py (hunk 1 above)
Comment (white238) [reply var/spack/repos/builtin/packages/axom/package.py]: Sounds good.  I removed the preferred all together. Am i right to assume that Spack will pick the highest numbered version if one isn't specified by the user?

File: var/spack/repos/builtin/packages/raja/package.py (hunk 3 above)
Comment (white238) [reply var/spack/repos/builtin/packages/raja/package.py:18]: No it was a copy and paste error.  I added that version back.

File: var/spack/repos/builtin/packages/axom/package.py (hunk 2 above)
Comment (white238) [reply var/spack/repos/builtin/packages/axom/package.py:159]: This is an example of how messy all of these options can be.  It would be a huge problem to add them to one command line. https://github.com/LLNL/axom/blob/develop/host-configs/rzansel-blueos_3_ppc64le_ib_p9-clang%408.0.1_nvcc_xlf.cmake

File: var/spack/repos/builtin/packages/axom/package.py [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+
+    path = os.path.realpath(path)
+
+    for key in path_replacements:
+        path = path.replace(key, path_replacements[key])
+
+    return path
+
+
+class Axom(Package):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/axom/package.py]: You should probably extend `CMakePackage` (since that's the build system you use), and `CudaPackage` (since you need to know what CUDA arch is supported.

File: var/spack/repos/builtin/packages/axom/package.py [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+            devtools_root = os.path.commonprefix([path1, path2])[:-1]
+            path_replacements[devtools_root] = "${DEVTOOLS_ROOT}"
+            cfg.write("# Root directory for generated developer tools\n")
+            cfg.write(cmake_cache_entry("DEVTOOLS_ROOT", devtools_root))
+
+        if "python" in spec or "devtools" in spec:
+            python_bin_dir = get_spec_path(spec, "python",
+                                           path_replacements, use_bin=True)
+            cfg.write(cmake_cache_entry("PYTHON_EXECUTABLE",
+                                        pjoin(python_bin_dir, "python")))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/axom/package.py]: This is only true for Python 2. For Python 3, the `python` executable may or may not exist. You should instead use `spec['python'].command.path`

File: var/spack/repos/builtin/packages/axom/package.py [hunk 6]
```diff
@@ -0,0 +1,497 @@
...
+        # Other machine specifics
+        ##################################
+
+        cfg.write("#------------------{0}\n".format("-" * 60))
+        cfg.write("# Other machine specifics\n")
+        cfg.write("#------------------{0}\n\n".format("-" * 60))
+
+        # OpenMP
+        if "+openmp" in spec:
+            cfg.write(cmake_cache_option("ENABLE_OPENMP", True))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/axom/package.py:402]: Can this be explicitly disabled in an else clause?

File: var/spack/repos/builtin/packages/py-shroud/package.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    version('0.10.1', tag='v0.10.1')
+    version('0.9.0', tag='v0.9.0')
+    version('0.8.0', tag='v0.8.0')
+
+    extends('python')
+
+    depends_on("py-alabaster")
+    depends_on("py-pytz")
+    depends_on("py-docutils")
+    depends_on("py-setuptools")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-shroud/package.py]:
```suggestion
    depends_on("py-setuptools", type='build')
```

File: var/spack/repos/builtin/packages/py-shroud/package.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    depends_on("py-pyyaml")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-shroud/package.py]:
```suggestion
    depends_on("py-pyyaml@4.2b1:", type=('build', 'run'))
```

File: var/spack/repos/builtin/packages/py-shroud/package.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    version('0.11.0', tag='v0.11.0')
+    version('0.10.1', tag='v0.10.1')
+    version('0.9.0', tag='v0.9.0')
+    version('0.8.0', tag='v0.8.0')
+
+    extends('python')
+
+    depends_on("py-alabaster")
+    depends_on("py-pytz")
+    depends_on("py-docutils")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-shroud/package.py]: These 3 dependencies aren't mentioned in the `setup.py`

File: var/spack/repos/builtin/packages/py-shroud/package.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack import *
+from spack.hooks.sbang import filter_shebang
+
+
+class PyShroud(Package):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-shroud/package.py]:
```suggestion
class PyShroud(PythonPackage):
```

File: var/spack/repos/builtin/packages/py-shroud/package.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    git      = "https://github.com/LLNL/shroud.git"
+
+    version('develop', branch='develop')
+    version('master',  branch='master')
+    version('0.11.0', tag='v0.11.0')
+    version('0.10.1', tag='v0.10.1')
+    version('0.9.0', tag='v0.9.0')
+    version('0.8.0', tag='v0.8.0')
+
+    extends('python')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-shroud/package.py]: Delete this line after you extend PythonPackage

File: var/spack/repos/builtin/packages/py-shroud/package.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+    depends_on("py-alabaster")
+    depends_on("py-pytz")
+    depends_on("py-docutils")
+    depends_on("py-setuptools")
+    depends_on("py-pyyaml")
+
+    def install(self, spec, prefix):
+        # simply install to the spack python
+        python('setup.py', 'install')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-shroud/package.py]: We never want to install directly to the Python installation prefix. It makes it impossible to uninstall a package, or install multiple versions of the same package. I would remove this whole `install` method.

File: var/spack/repos/builtin/packages/raja/package.py [hunk 13]
```diff
@@ -47,4 +48,13 @@ def cmake_args(self):
...
                 '-DCUDA_TOOLKIT_ROOT_DIR=%s' % (spec['cuda'].prefix)])
 
+        # Work around spack adding -march=ppc64le to SPACK_TARGET_ARGS which
+        # is used by the spack compiler wrapper.  This can go away when BLT
+        # removes -Werror from GTest flags
+        sys_type = ""
+        if "SYS_TYPE" in env:
+            sys_type = env["SYS_TYPE"]
+        if "blueos" in sys_type:
+            options.extend(['-DENABLE_TESTS=OFF'])
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/raja/package.py]:
```suggestion
            options.append('-DENABLE_TESTS=OFF')
```

File: var/spack/repos/builtin/packages/umpire/package.py [hunk 14]
```diff
@@ -68,4 +70,7 @@ def cmake_args(self):
         if '+numa' in spec:
             options.append('-DENABLE_NUMA=On')
 
+        if '+openmp' in spec:
+            options.append('-DENABLE_OPENMP=On')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/umpire/package.py]: Can you explicitly disable this in an else clause? Same for other variants.

File: var/spack/repos/builtin/packages/axom/package.py (hunk 1 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/axom/package.py]: You are correct

**Comment by white238** [comment]
I realize py-shroud disappeared for some reason during my rebase.  I'll fix it after confirming the last set of changes still work for our project.

**Comment by white238** [comment]
This passes all of Axom's internal testing and is good to be merged. Thanks for the help @becker33 and @adamjstewart !

**Comment by cyrush** [comment]
awesome work @white238 !
